components/servo/generic/servo.go

```diff
@@ -0,0 +1,337 @@
+			return nil, viamutils.NewConfigValidationError(path, errors.New("starting_position_degs cannot be lower than min_angle_deg"))
+		}
+		if config.MaxDeg != nil && *config.StartPos > *config.MaxDeg {
+			return nil, viamutils.NewConfigValidationError(path, errors.New("starting_position_degs cannot be higher than max_angle_deg"))
+		}
+		if *config.StartPos < minDeg || *config.StartPos > maxDeg {
+			return nil, viamutils.NewConfigValidationError(path, errors.New("starting_position_degs should be between 0 and 180"))
+		}
+	}
+	if config.MinDeg != nil && *config.MinDeg < minDeg {
+		return nil, viamutils.NewConfigValidationError(path, errors.Errorf("min_angle_deg cannot be lower than %f", minDeg))
+	}
+	if config.MaxDeg != nil && *config.MaxDeg > maxDeg {
+		return nil, viamutils.NewConfigValidationError(path, errors.Errorf("max_angle_deg cannot be higher than %f", maxDeg))
+	}
+	if config.MinWidthUS != nil && *config.MinWidthUS < minWidthUs {
+		return nil, viamutils.NewConfigValidationError(path, errors.Errorf("min_width_us cannot be lower than %d", minWidthUs))
+	}
+	if config.MaxWidthUS != nil && *config.MaxWidthUS > maxWidthUs {
+		return nil, viamutils.NewConfigValidationError(path, errors.Errorf("max_width_us cannot be higher than %d", maxWidthUs))
+	}
+	return deps, nil
+}
+
+const model = "generic"
+
+func init() {
+	registry.RegisterComponent(servo.Subtype, model,
+		registry.Component{
+			Constructor: newGenericServo,
+		})
+	config.RegisterComponentAttributeMapConverter(servo.SubtypeName, model,
+		func(attributes config.AttributeMap) (interface{}, error) {
+			var attr servoConfig
+			return config.TransformAttributeMapToStruct(&attr, attributes)
+		},
+		&servoConfig{})
+}
+
+type servoGeneric struct {
+	generic.Unimplemented
+	pin       board.GPIOPin
+	min       float64
+	max       float64
+	logger    golog.Logger
+	opMgr     operation.SingleOperationManager
+	frequency uint
+	minUs     uint
+	maxUs     uint
+	pwmRes    uint
+	currPct   float64
+}
+
+func newGenericServo(ctx context.Context, deps registry.Dependencies, cfg config.Component, logger golog.Logger) (interface{}, error) {
+	logger.Info("building servo!!")
+	attr, ok := cfg.ConvertedAttributes.(*servoConfig)
+	if !ok {
+		return nil, utils.NewUnexpectedTypeError(&servoConfig{}, cfg.ConvertedAttributes)
+	}
+	boardName := attr.Board
+	b, err := board.FromDependencies(deps, boardName)
+	if err != nil {
+		return nil, errors.Wrap(err, "board doesn't exist")
+	}
+	pin, err := b.GPIOPinByName(attr.Pin)
+	if err != nil {
+		return nil, errors.Wrap(err, "couldn't get servo pin")
+	}
+	frequency, err := pin.PWMFreq(ctx, nil)
+	if err != nil {
+		return nil, errors.Wrap(err, "couldn't get servo pin pwm frequency")
+	}
+	if attr.Frequency != nil {
+		err = pin.SetPWMFreq(ctx, *attr.Frequency, nil)
+		if err != nil {
+			return nil, errors.Wrap(err, "error setting servo pin frequency")
+		}
+		frequency = *attr.Frequency
+	}
+	if frequency > 500 {
```

Comment:
To double-check: this should not be `if frequency > minWidthUs`, right? The numbers are the same, but the units seem different (hertz vs. microseconds).

Reply:
if the frequency is above 500hz it's impossible to encode a 2ms pulse